### ISSUE TITLE
Explain Haskell/Java performance comparison

### ISSUE THREAD

**Issue by limick** [Gabriel439/post-rfc#108]
`Haskell's performance is excellent and currently comparable to Java. Both languages give roughly the same performance in beginner or expert hands, although for different reasons.`

I have a hard time understanding what this means. Would you mind explaining why this is the case, i.e. listing the reasons?

**Comment by Gabriella439** [comment]
I will explain my reasoning.  Note that I know much more about Haskell performance than Java performance and there is a good chance that anything I say about Java is incorrect, so please correct me.

Here is my understanding of the pitfalls that affect beginners in each language:

* For Haskell the pitfalls that most commonly affect beginners are:
    * Poor defaults for data structures in the standard library (i.e. `String` and linked lists)
    * Difficult to reason about performance due to laziness
* For Java the pitfalls that most commonly affect beginners are:
    * Poor defaults for multithreaded programming, serialization, and runtime options
    * Difficult to reason about performance due to JIT optimization
* Pitfalls that they share in common (for beginners)
    * Garbage collection
    * Lack of cache coherency

Here is my understanding of how experts achieve high performance in each language:

* Haskell performance tricks
    * Take precautions to ensure that shortcut fusion / deforestation works correctly
    * Encode recursive data types non-recursively to improve optimization
    * Take advantage of unboxed and cache-coherent data structures
    * Reduce the number of objects on the heap to reduce GC pressure
        * Note: This is good to do for Java, too, but my unscientific impression is that this is even more critical for Haskell's garbage collector
    * Write "C in Haskell" (i.e. use the `Foreign.*` module hierarchy)
* Java performance tricks
    * Profile extensively
        * Note: This is good to do for Haskell, too, but yields more dividends in Java
    * Use thread pools and asynchronous IO
    * Use long-lived objects and/or memory pools to reduce GC pressure
* Both:
    * Tune runtime options
    * Use the right algorithms and data structures

That's my rough first pass.  Feel free to comment or criticize that

**Comment by JulianLeviston** [comment]
@Gabriel439 this is a pretty nice beginning of a summary, and it'd be nice if it was either in the SOTU, or referenced from it somehow (blog post?)

**Comment by limick** [comment]
That's a great answer and I agree that it deserves to be published somewhere.
In the meantime, I would maybe amend that second sentence to something like

> Both languages give roughly the same performance in beginner or expert hands, although tuning strategies differ considerably.

**Comment by JulianLeviston** [comment]
Personally I think the way it's written is fine as is.

**Comment by limick** [comment]
I found that sentence hard to parse, and I think it wouldn't hurt to spell things out a bit, but all in all it's a very minor issue.
If more people have trouble understanding, the issue will come up again.

**Comment by JulianLeviston** [comment]
@limick I'm still very interested in how it was hard to parse. (Language, understanding and learning are my absolute favourite topics, and this hits all three.)

**Comment by limick** [comment]
@JulianLeviston I had two issues with the sentence:

> Both languages give roughly the same performance in beginner or expert hands, although for different reasons.

1. Grammatically speaking, I think it could be interpreted as "performance of Java in beginner hands == performance of Java in expert hands" (and same for Haskell). Of course it doesn't take a genius to figure out that these statements are not true, and therefore that's not the intended meaning, but I think such an ambiguous sentence structure should be avoided to begin with. A bit like we try to make illegal states unrepresentable in code.
2. The sentence (as explained by Gabriel) contains a lot of information. Unrolling and paraphrasing a bit:
    - The performance of Haskell in beginner hands is roughly the same as the performance of Java in beginner hands.
    - The performance of Haskell in expert hands is roughly the same as the performance of Java in expert hands.
    - The reasons why Haskell is slow in beginner hands are different from the reasons why Java is slow in beginner hands (i.e. pitfalls are different).
    - The reasons why Haskell is fast in expert hands are different from the reasons why Java is fast in expert hands (i.e. tuning strategies are different).

    Even after the attentive reader mentally unrolled these four points, they still don't know what the different reasons are.

Again, in the context of the document it's a minor point and it may be the wrong place to expand on the question of performance. That's why I think it may be best to simplify the sentence a bit and refer to a more detailed document somewhere else.
I hope that clarified it a bit.

**Comment by JulianLeviston** [comment]
That's a really cool answer. Thank you very much! I agree.
